lib/AFP/Data/ByteString/Short.hs

```diff
@@ -0,0 +1,676 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE MultiWayIf #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE UnboxedTuples #-}
+
+module AFP.Data.ByteString.Short
```

Comment:
Basically this is `Data.ByteString.Short` on steroids, with many functions added.

We re-export the following:

```hs
import Data.ByteString.Short
    ( ShortByteString
    , empty
    , fromShort
    , index
    , length
    , null
    , pack
    , packCString
    , packCStringLen
    , toShort
    , unpack
    , useAsCString
    , useAsCStringLen
    )
```

The rest is written here and needs explicit review and also more tests and benchmarks. Also check the complexity and the copying behavior.

Many functions have to perform some form of memory copy.